Controller gRPC server is rejecting client keep-alives

**Problem description**
The gRPC server is rejected client keep-alive pings with the following exceptions, hence all gRPC calls taking longer than 1 minute fails:

Jul 20, 2017 11:02:42 AM io.grpc.netty.NettyClientHandler$1 onGoAwayReceived
WARNING: Received GOAWAY with ENHANCE_YOUR_CALM. Debug data: {1}
Jul 20, 2017 11:02:42 AM io.grpc.internal.AtomicBackoff$State backoff
WARNING: Increased keepalive time nanos to 120,000,000,000
2017-07-20 11:02:42,709 61492 [grpc-default-executor-0] WARN  i.p.c.stream.impl.ControllerImpl - gRPC call failed with server error: RESOURCE_EXHAUSTED: Bandwidth exhausted
HTTP/2 error code: ENHANCE_YOUR_CALM
Received Goaway
too_many_pings
2017-07-20 11:02:42,710 61493 [grpc-default-executor-0] WARN  i.p.c.stream.impl.ControllerImpl - scaleStream failed: 
java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Bandwidth exhausted
HTTP/2 error code: ENHANCE_YOUR_CALM
Received Goaway
too_many_pings
    at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
    at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
    at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:593)
    at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
    at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
    at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
    at io.pravega.client.stream.impl.ControllerImpl$RPCAsyncCallback.onError(ControllerImpl.java:660)
    at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:395)
    at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
    at io.grpc.internal.ClientCallImpl.access$100(ClientCallImpl.java:76)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:512)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:429)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:544)
    at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:52)
    at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:117)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Bandwidth exhausted
HTTP/2 error code: ENHANCE_YOUR_CALM
Received Goaway
too_many_pings
    at io.grpc.Status.asRuntimeException(Status.java:543)
    ... 11 common frames omitted
2017-07-20 11:02:42,710 61493 [grpc-default-executor-0] TRACE i.p.c.stream.impl.ControllerImpl - LEAVE scaleStream@3186427135754 (elapsed=59805415us).

**Problem location**
ControllerImpl

**Suggestions for an improvement**
Increase the keep-alive timeout to be within the gRPC server's permissible value.

--- a/client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java
+++ b/client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java
@@ -77,7 +77,8 @@ import static io.pravega.common.concurrent.FutureHelpers.getAndHandleExceptions;
 public class ControllerImpl implements Controller {
 
     // The default keepalive interval for the gRPC transport to ensure long running RPCs are tested for connectivity.
-    private static final long DEFAULT_KEEPALIVE_TIME_SECONDS = 60;
+    // This value should be greater than the permissible value configured at the server which is by default 5 minutes.
+    private static final long DEFAULT_KEEPALIVE_TIME_MINUTES = 6;
 
     // The gRPC client for the Controller Service.
     private final ControllerServiceGrpc.ControllerServiceStub client;
@@ -95,7 +96,7 @@ public class ControllerImpl implements Controller {
         this(NettyChannelBuilder.forTarget(controllerURI.toString())
                 .nameResolverFactory(new ControllerResolverFactory())
                 .loadBalancerFactory(RoundRobinLoadBalancerFactory.getInstance())
-                .keepAliveTime(DEFAULT_KEEPALIVE_TIME_SECONDS, TimeUnit.SECONDS)
+                .keepAliveTime(DEFAULT_KEEPALIVE_TIME_MINUTES, TimeUnit.MINUTES)
                 .usePlaintext(true));
         log.info("Controller client connecting to server at {}", controllerURI.getAuthority());
     }

--- a/client/src/test/java/io/pravega/client/stream/impl/ControllerImplTest.java
+++ b/client/src/test/java/io/pravega/client/stream/impl/ControllerImplTest.java
@@ -11,15 +11,17 @@ package io.pravega.client.stream.impl;
 
 import com.google.common.collect.ImmutableSet;
 import io.grpc.Status;
-import io.grpc.inprocess.InProcessChannelBuilder;
-import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.StatusRuntimeException;
 import io.grpc.internal.ServerImpl;
+import io.grpc.netty.NettyChannelBuilder;
+import io.grpc.netty.NettyServerBuilder;
 import io.grpc.stub.StreamObserver;
 import io.pravega.client.segment.impl.Segment;
 import io.pravega.client.stream.ScalingPolicy;
 import io.pravega.client.stream.Stream;
 import io.pravega.client.stream.StreamConfiguration;
 import io.pravega.client.stream.Transaction;
+import io.pravega.common.Exceptions;
 import io.pravega.controller.stream.api.grpc.v1.Controller;
 import io.pravega.controller.stream.api.grpc.v1.Controller.CreateScopeStatus;
 import io.pravega.controller.stream.api.grpc.v1.Controller.CreateStreamStatus;
@@ -49,6 +51,7 @@ import io.pravega.controller.stream.api.grpc.v1.ControllerServiceGrpc.Controller
 import io.pravega.shared.protocol.netty.PravegaNodeUri;
 import io.pravega.test.common.AssertExtensions;
 import java.io.IOException;
+import java.net.URI;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
@@ -57,11 +60,14 @@ import java.util.Map.Entry;
 import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Semaphore;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+
+import io.pravega.test.common.TestUtils;
 import lombok.val;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.lang3.tuple.Pair;
@@ -84,10 +90,13 @@ public class ControllerImplTest {
     private static final int SERVICE_PORT = 12345;
 
     @Rule
-    public final Timeout globalTimeout = new Timeout(20, TimeUnit.SECONDS);
+    public final Timeout globalTimeout = new Timeout(120, TimeUnit.SECONDS);
 
     // Test implementation for simulating the server responses.
-    private ServerImpl fakeServer = null;
+    private ControllerServiceImplBase testServerImpl;
+    private ServerImpl testGRPCServer = null;
+
+    private int serverPort;
 
     // The controller RPC client.
     private ControllerImpl controllerClient = null;
@@ -95,8 +104,8 @@ public class ControllerImplTest {
     @Before
     public void setup() throws IOException {
 
-        // Setup fake server generating different success and failure responses.
-        ControllerServiceImplBase fakeServerImpl = new ControllerServiceImplBase() {
+        // Setup test server generating different success and failure responses.
+        testServerImpl = new ControllerServiceImplBase() {
             @Override
             public void createStream(StreamConfig request,
                     StreamObserver<CreateStreamStatus> responseObserver) {
@@ -128,16 +137,19 @@ public class ControllerImplTest {
                 } else if (request.getStreamInfo().getStream().equals("streamparallel")) {
 
                     // Simulating delay in sending response.
-                    try {
-                        Thread.sleep(500);
-                    } catch (InterruptedException e) {
-                        log.error("Unexpected interrupt");
-                        responseObserver.onError(e);
-                        return;
-                    }
+                    Exceptions.handleInterrupted(() -> Thread.sleep(500));
                     responseObserver.onNext(CreateStreamStatus.newBuilder()
                                                     .setStatus(CreateStreamStatus.Status.SUCCESS)
                                                     .build());
+                    responseObserver.onCompleted();
+                } else if (request.getStreamInfo().getStream().equals("streamdelayed")) {
+
+                    // Simulating delay in sending response. This is used for the keepalive test,
+                    // where response time > 30 seconds is required to simulate a failure.
+                    Exceptions.handleInterrupted(() -> Thread.sleep(40000));
+                    responseObserver.onNext(CreateStreamStatus.newBuilder()
+                            .setStatus(CreateStreamStatus.Status.SUCCESS)
+                            .build());
                     responseObserver.onCompleted();
                 } else {
                     responseObserver.onError(Status.INTERNAL.withDescription("Server error").asRuntimeException());
@@ -250,13 +262,7 @@ public class ControllerImplTest {
                                                     .build());
                     responseObserver.onCompleted();
                 } else if (request.getStream().equals("streamparallel")) {
-                    try {
-                        Thread.sleep(500);
-                    } catch (InterruptedException e) {
-                        log.error("Unexpected interrupt");
-                        responseObserver.onError(e);
-                        return;
-                    }
+                    Exceptions.handleInterrupted(() -> Thread.sleep(500));
                     responseObserver.onNext(SegmentRanges.newBuilder()
                                                     .addSegmentRanges(ModelHelper.createSegmentRange("scope1",
                                                                                                      "streamparallel",
@@ -539,17 +545,49 @@ public class ControllerImplTest {
             }
         };
 
-        fakeServer = InProcessServerBuilder.forName("fakeserver")
-                .addService(fakeServerImpl)
-                .directExecutor()
+        serverPort = TestUtils.getAvailableListenPort();
+        testGRPCServer = NettyServerBuilder.forPort(serverPort)
+                .addService(testServerImpl)
                 .build()
                 .start();
-        controllerClient = new ControllerImpl(InProcessChannelBuilder.forName("fakeserver").directExecutor());
+        controllerClient = new ControllerImpl(URI.create("tcp://localhost:" + serverPort));
     }
 
     @After
     public void tearDown() {
-        fakeServer.shutdown();
+        testGRPCServer.shutdownNow();
+    }
+
+    @Test
+    public void testKeepAlive() throws IOException, ExecutionException, InterruptedException {
+
+        // Verify that keep-alive timeout less than permissible by the server results in a failure.
+        ControllerImpl controller = new ControllerImpl(NettyChannelBuilder.forAddress("localhost", serverPort)
+                .keepAliveTime(10, TimeUnit.SECONDS).usePlaintext(true));
+        CompletableFuture<Boolean> createStreamStatus = controller.createStream(StreamConfiguration.builder()
+                .streamName("streamdelayed")
+                .scope("scope1")
+                .scalingPolicy(ScalingPolicy.fixed(1))
+                .build());
+        AssertExtensions.assertThrows("Should throw Exception", createStreamStatus,
+                throwable -> throwable instanceof StatusRuntimeException);
+
+        // Verify that the same RPC with permissible keepalive time succeeds.
+        int serverPort2 = TestUtils.getAvailableListenPort();
+        ServerImpl testServer = NettyServerBuilder.forPort(serverPort2)
+                .addService(testServerImpl)
+                .permitKeepAliveTime(5, TimeUnit.SECONDS)
+                .build()
+                .start();
+        controller = new ControllerImpl(NettyChannelBuilder.forAddress("localhost", serverPort2)
+                .keepAliveTime(10, TimeUnit.SECONDS).usePlaintext(true));
+        createStreamStatus = controller.createStream(StreamConfiguration.builder()
+                .streamName("streamdelayed")
+                .scope("scope1")
+                .scalingPolicy(ScalingPolicy.fixed(1))
+                .build());
+        assertTrue(createStreamStatus.get());
+        testServer.shutdownNow();
     }
 
     @Test

